### PR TITLE
Glharper/speaker recognition

### PR DESCRIPTION
**Pull request by glharper**
Add speaker recognition feature, with primary flows going through VoiceProfileClient and SpeakerRecognizer classes.

### CODE REVIEW

File: src/common.browser/FileAudioSource.ts [hunk 1]
```diff
@@ -56,6 +56,12 @@ export class FileAudioSource implements IAudioSource {
         return this.privAudioFormatPromise;
     }
 
+    public get blob(): Promise<Blob> {
+        const blobDeferral: Deferred<Blob> = new Deferred<Blob>();
+        blobDeferral.resolve(this.privFile);
+        return blobDeferral.promise();
```
Comment (rhurey) [inline src/common.browser/FileAudioSource.ts:62]: You should be able to use PromiseHelper.fromResult here.

Comment (glharper) [reply src/common.browser/FileAudioSource.ts]: Corrected.

File: src/common.browser/FileAudioSource.ts [hunk 2]
```diff
@@ -141,46 +145,48 @@ export class FileAudioSource implements IAudioSource {
...
     private readHeader = (): Promise<AudioStreamFormatImpl> => {
         // Read the wave header.
-        const header: Blob = this.privFile.slice(0, 44);
-        const headerReader: FileReader = new FileReader();
-
         const headerResult: Deferred<AudioStreamFormatImpl> = new Deferred<AudioStreamFormatImpl>();
 
-        const processHeader = (event: Event): void => {
-            const header: ArrayBuffer = (event.target as FileReader).result as ArrayBuffer;
+        if (typeof (FileReader) !== "undefined") {
```
Comment (rhurey) [inline src/common.browser/FileAudioSource.ts:150]: >if (typeof (FileReader) !== "undefined") [](start = 8, length = 40)

Where did this happen, and if we can't read the header won't this just leave the promise permanently not resolved? Should it be rejected?

File: src/common.browser/MicAudioSource.ts [hunk 3]
```diff
@@ -78,6 +78,10 @@ export class MicAudioSource implements IAudioSource {
         return PromiseHelper.fromResult(MicAudioSource.AUDIOFORMAT);
     }
 
+    public get blob(): Promise<Blob> {
+        throw new Error("Mic input has no blob attached");
```
Comment (rhurey) [inline src/common.browser/MicAudioSource.ts:82]: >Error [](start = 18, length = 5)

Why not reject the promise? Not sure we should mix error styles.

File: src/common.browser/RestConfigBase.ts [hunk 4]
```diff
@@ -0,0 +1,81 @@
+
+import { IErrorMessages } from "../common/Exports";
```
Comment (rhurey) [inline src/common.browser/RestConfigBase.ts:2]: Needs the copyright header.

File: src/common.browser/RestMessageAdapter.ts [hunk 5]
```diff
@@ -0,0 +1,279 @@
...
+    private privSendQueue: Queue<ISendItem>;
+
+    public constructor(
+        configParams: IRequestOptions,
+        connectionId?: string
+        ) {
+
+        if (!configParams) {
+            throw new ArgumentNullError("configParams");
+        }
```
Comment (rhurey) [inline src/common.browser/RestMessageAdapter.ts:66]: Look for Contracts.ThrowIfNull

File: src/common.browser/RestMessageAdapter.ts [hunk 6]
```diff
@@ -0,0 +1,279 @@
...
+    public open(): Promise<ConnectionOpenResponse> {
+        const responseReceivedDeferral = new Deferred<ConnectionOpenResponse>();
+        const ontimeout = () => {
+            responseReceivedDeferral.resolve(new ConnectionOpenResponse(200, ""));
+        };
+        this.privSendQueue = new Queue<ISendItem>();
+        setTimeout(ontimeout, 20);
+        this.processSendQueue();
+        return responseReceivedDeferral.promise();
+    }
```
Comment (rhurey) [inline src/common.browser/RestMessageAdapter.ts:88]: Why does a REST adapter that's not going to persist the connection need an open() method?

File: src/sdk/SpeakerIdentificationModel.ts [hunk 7]
```diff
@@ -0,0 +1,33 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT license.
+
+import { Contracts } from "./Contracts";
+import {
+    VoiceProfile,
+    VoiceProfileType,
+} from "./Exports";
+
+export class SpeakerIdentificationModel {
```
Comment (rhurey) [inline src/sdk/SpeakerIdentificationModel.ts:10]: Needs doc tags

File: src/sdk/SpeakerIdentificationModel.ts [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+    private privVoiceProfiles: VoiceProfile[] = [];
+
+    private constructor(profiles: VoiceProfile[]) {
+        Contracts.throwIfNullOrUndefined(profiles, "VoiceProfiles");
+        if (profiles.length === 0) {
+            throw new Error("Empty Voice Profiles array");
+        }
+        profiles.forEach((profile: VoiceProfile) => {
+            if (profile.profileType !== VoiceProfileType.TextIndependentIdentification) {
+                throw new Error("Identification model cannot be created from Verification profile: " + profile.profileId);
```
Comment (rhurey) [inline src/sdk/SpeakerIdentificationModel.ts:20]: >Verification [](start = 77, length = 12)

I'd rephrase as "Can only be made from...." because if another profile type appears the message will be misleading.

File: src/sdk/VoiceProfileResult.ts [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+        return this.privProperties;
+    }
+
+    public get errorDetails(): string {
+        return this.privErrorDetails;
+    }
+}
+
+/**
+ * Output format
```
Comment (rhurey) [inline src/sdk/VoiceProfileResult.ts:47]: >Output format [](start = 3, length = 13)

?

File: src/sdk/SpeakerRecognizer.ts [hunk 10]
```diff
@@ -0,0 +1,165 @@
...
+     * @returns {PropertyCollection} The collection of properties and their values defined for this SpeakerRecognizer.
+     */
+    public get properties(): PropertyCollection {
+        return this.privProperties;
+    }
+
+    /**
+     * SpeakerRecognizer constructor.
+     * @constructor
+     * @param {SpeechConfig} speechConfig - An set of initial properties for this synthesizer (authentication key, region, &c)
```
Comment (rhurey) [inline src/sdk/SpeakerRecognizer.ts:75]: >synthesizer [](start = 82, length = 11)

?

File: src/sdk/VoiceProfileClient.ts [hunk 11]
```diff
@@ -0,0 +1,238 @@
...
+     * @function
+     * @public
+     * @param {VoiceProfileType} profileType Type of Voice Profile to be created
+     *        specifies the keyword to be recognized.
+     * @param {string} lang Language string (locale) for Voice Profile
+     * @param cb - Callback invoked once Voice Profile has been created.
+     * @param err - Callback invoked in case of an error.
+     */
+    public createProfileAsync(profileType: VoiceProfileType, lang: string, cb?: (e: VoiceProfile) => void, err?: (e: string) => void): void {
+        this.privAdapter.createProfile(profileType, lang).continueWith((promiseResult: PromiseResult<IRestResponse>) => {
```
Comment (rhurey) [inline src/sdk/VoiceProfileClient.ts:96]: >continueWith [](start = 58, length = 12)

Why not call promise.on here instead of getting the result and using if statements?

File: src/sdk/Audio/AudioInputStream.ts [hunk 12]
```diff
@@ -346,6 +350,10 @@ export class PullAudioInputStreamImpl extends PullAudioInputStream implements IA
         return this.privId;
     }
 
+    public get blob(): Promise<Blob> {
+        return undefined;
```
Comment (rhurey) [inline src/sdk/Audio/AudioInputStream.ts:354]: Should this  be a rejected promise?

File: src/common.browser/FileAudioSource.ts (hunk 2 above)
Comment (glharper) [reply src/common.browser/FileAudioSource.ts]: Reverted to original flow

File: src/common.browser/RestMessageAdapter.ts (hunk 6 above)
Comment (glharper) [reply src/common.browser/RestMessageAdapter.ts]: This was in case I had to adapt to Session based connection style. Removed.

File: src/sdk/SpeakerIdentificationModel.ts (hunk 8 above)
Comment (glharper) [reply src/sdk/SpeakerIdentificationModel.ts]: Changed.

File: src/sdk/SpeakerRecognizer.ts (hunk 10 above)
Comment (glharper) [reply src/sdk/SpeakerRecognizer.ts]: corrected.

File: src/sdk/Audio/AudioInputStream.ts (hunk 12 above)
Comment (glharper) [reply src/sdk/Audio/AudioInputStream.ts]: Changed.

File: src/sdk/VoiceProfileClient.ts (hunk 11 above)
Comment (glharper) [reply src/sdk/VoiceProfileClient.ts:96]: Changed

File: src/sdk/VoiceProfileResult.ts (hunk 9 above)
Comment (glharper) [reply src/sdk/VoiceProfileResult.ts]: Removed

File: src/sdk/SpeakerIdentificationModel.ts (hunk 7 above)
Comment (glharper) [reply src/sdk/SpeakerIdentificationModel.ts:15]: Added for this and SpeakerVerificationModel

File: src/common.browser/RestMessageAdapter.ts (hunk 5 above)
Comment (glharper) [reply src/common.browser/RestMessageAdapter.ts:56]: Would rather not add a dependency on sdk module if it can be avoided.

File: src/common.browser/RestConfigBase.ts (hunk 4 above)
Comment (glharper) [reply src/common.browser/RestConfigBase.ts:4]: Added

File: src/common.browser/MicAudioSource.ts (hunk 3 above)
Comment (glharper) [reply src/common.browser/MicAudioSource.ts]: Changed to reject promise

File: src/common.speech/SpeakerIdMessageAdapter.ts [hunk 13]
```diff
@@ -0,0 +1,162 @@
...
+    private privRestAdapter: RestMessageAdapter;
+    private privUri: string;
+
+    public constructor(config: SpeakerRecognitionConfig) {
+
+        const connectionId: string = createNoDashGuid();
+        let endpoint = config.parameters.getProperty(PropertyId.SpeechServiceConnection_Endpoint, undefined);
+        if (!endpoint) {
+            const region: string = config.parameters.getProperty(PropertyId.SpeechServiceConnection_Region, "westus");
+            const host: string = config.parameters.getProperty(PropertyId.SpeechServiceConnection_Host, "https://" + region + ".api.cognitive.microsoft.com/speaker/{mode}/v2.0/{dependency}");
```
Comment (yulin-li) [inline src/common.speech/SpeakerIdMessageAdapter.ts:39]: although speaker recognition is not available in azure.cn for now, it's better to support azure.cn regions here. You can refer [here](https://github.com/microsoft/cognitive-services-speech-sdk-js/blob/master/src/common.speech/SpeechConnectionFactory.ts#L46)

Comment (glharper) [reply src/common.speech/SpeakerIdMessageAdapter.ts]: Updated to support azure.cn if chinese region detected

File: src/common.browser/MicAudioSource.ts [hunk 14]
```diff
@@ -78,6 +78,12 @@ export class MicAudioSource implements IAudioSource {
         return PromiseHelper.fromResult(MicAudioSource.AUDIOFORMAT);
     }
 
+    public get blob(): Promise<Blob> {
+        const deferred = new Deferred<Blob>();
+        deferred.reject("Not implemented for Mic input");
+        return deferred.promise();
```
Comment (rhurey) [inline src/common.browser/MicAudioSource.ts:84]: There should be a PromiseHelper.reject (or something close)

File: src/sdk/Audio/AudioInputStream.ts [hunk 15]
```diff
@@ -346,6 +383,12 @@ export class PullAudioInputStreamImpl extends PullAudioInputStream implements IA
         return this.privId;
     }
 
+    public get blob(): Promise<Blob | Buffer> {
+        const deferred = new Deferred<Blob>();
+        deferred.reject("Not implemented");
+        return deferred.promise();
```
Comment (rhurey) [inline src/sdk/Audio/AudioInputStream.ts:389]: PromiseHelper

**Comment by rhurey** [comment]
>     public static VoiceSignatureEnrollmentEndpoint: string = "wss://10.91.185.105:8446/speech/voicesignature?&language=en-us";

Needs to change to a prod endpoint.

---
Refers to: tests/Settings.ts:11 in a726a61. [](commit_id = a726a61ed6756d5145d756d9b2a7ecd080caa221, deletion_comment = False)

**Comment by rhurey** [comment]
The only work that's important is to get the tests passing against a prod endpoint, the yml and ADO pipeline will need to be updated to use WestUS for now.

File: src/common.browser/MicAudioSource.ts (hunk 14 above)
Comment (glharper) [reply src/common.browser/MicAudioSource.ts:84]: PromiseHelper.fromError, it looks like

Comment (glharper) [reply src/common.browser/MicAudioSource.ts]: Corrected

File: src/sdk/Audio/AudioInputStream.ts (hunk 15 above)
Comment (glharper) [reply src/sdk/Audio/AudioInputStream.ts]: Corrected
